File: 121.best-time-to-buy-and-sell-stock.py
#
# @lc app=leetcode id=121 lang=python3
#
# [121] Best Time to Buy and Sell Stock
#

"""
maxProfit

"""

# @lc code=start
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_buy_price, max_profit = prices[0], 0
        for price in prices[1:]:
            if price < min_buy_price:
                min_buy_price = price
            if price - min_buy_price > max_profit:
                max_profit = price - min_buy_price

        return max_profit
# @lc code=end

File: Rust/121.best-time-to-buy-and-sell-stock.rs
/*
 * @lc app=leetcode id=121 lang=rust
 *
 * [121] Best Time to Buy and Sell Stock
 */

// @lc code=start
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut min_buy_price = i32::MAX;
        let mut max_profit = 0;

        for i in 0..prices.len() - 1 {
            min_buy_price = i32::min(min_buy_price, prices[i]);

            if prices[i + 1] - min_buy_price > max_profit {
                max_profit = prices[i + 1] - min_buy_price;
            }
        }

        max_profit
    }
}
// @lc code=end
